File: code/example/javascript/wave.js
function wave(str) {
    let wave = [];
    // Convert original string to lowercase
    let strLower = str.toLowerCase();

    // Loop through the characters of the string
    for (let i = 0; i < str.length; i++) {
        // Check if current character is letter using regular expression
        if (/[a-zA-Z]/.test(str[i])) {
            // Create substring by concatenating the sliced lowercase string from beginning until current index,
            // uppercase current character and sliced lowercase string after the index of the current character until end
            let substr = strLower.slice(0, i) + str[i].toUpperCase() + strLower.slice(i + 1);
            // Push at the end of result list of strings
            wave.push(substr);
        }
    }

    console.log(wave);
    return wave;
}

// Call the function with one input argument
if (process.argv.length > 2) {
    wave(process.argv[2]);
} else {
    console.log("No input string provided.");
}

File: code/example/java/Wave.java
package code.example.java;

import java.util.ArrayList;

public class Wave {
    public static void main(String[] args){
        ArrayList<String> wave = new ArrayList<>();

        // Read first argument as the input string
        String str = "";
        if (args.length == 0) {
            System.out.println("No input string provided.");
        } else {
            str = args[0];
        }

        // Use StringBuilder to reduce complexity from lowercase string
        StringBuilder strBuilder = new StringBuilder(str.toLowerCase());

        // Loop through the characters of the string
        for (int i = 0; i < str.length(); i++) {
            // Find character at index i
            char c = str.charAt(i);
            // Check if character is letter
            if (Character.isLetter(c)) {
                // Set character at index i (current) to uppercase
                strBuilder.setCharAt(i, Character.toUpperCase(c));
                // Add the resulted string builder to result array
                wave.add(strBuilder.toString());
                // Revert the change of the current character back to lowercase
                strBuilder.setCharAt(i, Character.toLowerCase(c));
            }
        }

        System.out.println(wave);
    }
}
